avail/src/main/kotlin/avail/anvil/Styles.kt

```diff
@@ -32,605 +32,2987 @@
+	 * @return
+	 *   The combined [context][ValidatedRenderingContext].
 	 */
-	@Suppress("unused")
-	inline fun <reified T> addStyles(doc: StyledDocument)
-		where T: Enum<T>, T: BoundStyle
+	fun overrideWith(
+		other: ValidatedRenderingContext
+	): ValidatedRenderingContext
 	{
-		T::class.java.enumConstants!!.forEach { it.setStyle(doc) }
+		assert(palette === other.palette)
+		val a = attributes
+		val o = other.attributes
+		return ValidatedRenderingContext(
+			attributes.copy(
+				fontFamily = o.fontFamily ?: a.fontFamily,
+				foreground = o.foreground ?: a.foreground,
+				background = o.background ?: a.background,
+				bold = o.bold ?: a.bold,
+				italic = o.italic ?: a.italic,
+				underline = o.underline ?: a.underline,
+				superscript = o.superscript ?: a.superscript,
+				subscript = o.subscript ?: a.subscript,
+				strikethrough = o.strikethrough ?: a.strikethrough
+			),
+			palette
+		)
 	}
 
 	/**
-	 * Apply the appropriate variant of every registered style to the specified
-	 * [StyledDocument].
+	 * Apply the [receiver][RenderingContext] to the specified [range] of the
+	 * target [StyledDocument].
 	 *
-	 * @param doc
-	 *   The styled document.
+	 * @param document
+	 *   The target [StyledDocument].
+	 * @param classifiers
+	 *   The classifier sequence associated with the [range]. A
+	 *   [NameAttribute][StyleConstants.NameAttribute] will be attached to the
+	 *   [range], to support the style classifier introspection feature.
+	 * @param range
+	 *   The target one-based [range][IntRange] of characters within the
+	 *   [document].
+	 * @param replace
+	 *   Indicates whether or not the previous attributes should be cleared
+	 *   before the new attributes are set. If `true`, the operation will
+	 *   replace the previous attributes entirely. If `false`, the new
+	 *   attributes will be merged with the previous attributes.
 	 */
-	fun addAllStyles(doc: StyledDocument) = allStyles.values.forEach {
-		it.setStyle(doc)
+	fun renderTo(
+		document: StyledDocument,
+		classifiers: String,
+		range: IntRange,
+		replace: Boolean = true
+	)
+	{
+		assert(SwingUtilities.isEventDispatchThread())
+		// This line is subtle, because it binds `classifiers` to a
+		// NameAttribute, which in turn makes the classifiers available for
+		// display in gutter of the AvailEditor.
+		val style = document.addStyle(classifiers, defaultDocumentStyle)
+		style.addAttributes(documentAttributes)
+		document.setCharacterAttributes(
+			range.first - 1,
+			range.last - range.first + 1,
+			style,
+			replace)
 	}
 
-	override fun iterator() = allStyles.values.iterator()
+	override fun hashCode() = 31 * super.hashCode()
+
+	companion object
+	{
+		/**
+		 * The default [document&#32;style][Style], to serve as the parent for
+		 * new document styles. **Must only be called on the Swing UI thread.**
+		 */
+		val defaultDocumentStyle: Style
+		by lazy(LazyThreadSafetyMode.SYNCHRONIZED) {
+			assert(SwingUtilities.isEventDispatchThread())
```

Comment:
I'm not sure that this is true, because theoretically you're only supposed to manipulate these `StyleContext`s from the Swing UI thread. So it should be an error to query this field from some other thread.